### PR TITLE
test(fire): recursive-index benchmarks and published sizing guidance

### LINKED ISSUE #812: FIRE-20 — Performance benchmarks for recursive fire indices
**Labels:** `performance`, `fire-weather`, `tests`
**Milestone:** `v2.8 — Fire weather diagnostics`
**Depends on:** #799, #804

### Summary

`performance.py` already exists in the package. Extend it to cover the fire
family, because the recursive indices have a fundamentally different scaling
profile from SPI and SPEI and users need realistic expectations.

### Tasks

- [ ] Benchmark KBDI and full CFFWIS at several grid sizes and record lengths
- [ ] Compare pure NumPy against the optional accelerated kernel from #795
- [ ] Measure the single-pass orchestrator against separately chained calls
- [ ] Document peak memory as a function of chunk size, given time cannot be
      chunked
- [ ] Add a CI performance-regression guard consistent with existing practice
- [ ] Publish a sizing table in the docs: "computing CFFWIS over CONUS at 0.25°
      for 40 years takes approximately X on Y"

### Acceptance criteria

- Benchmark results committed and reproducible
- Sizing guidance published in the fire docs page
- Regression guard fails on a deliberate slowdown

### PR DESCRIPTION
**Pull request by monocongo**
Closes #812.

## What this adds

`tests/test_benchmark_fire.py` measures the fire family's scaling profile and
turns it into the sizing guidance the issue asks for, plus the regression guards.

- **Throughput** in cell-days per second for CFFWIS and KBDI across grid sizes
  and record lengths (`@pytest.mark.benchmark`, group `fire-scaling`).
- **Orchestrator cost**: `fire.cffwis()` in one pass against the separately
  chained `ffmc` / `duff_moisture_code` / `drought_code` calls and the derived
  indices (group `fire-orchestrator`).
- **Peak RSS** across spatial chunk sizes and output selection, measured in a
  fresh process per row (group `fire-memory`).
- **Guards** (group `fire-guards`): the benchmarks workflow runs them on every
  pull request and they fail once the recurrence costs about 1.7x more per
  cell-day, once the orchestrator costs more than 1.1x the chained calls it
  replaces, or once peak RSS leaves the coarse bound around the modeled
  footprint. Their failure paths are covered deterministically by
  `TestFireBudgetPolicy` in the default suite, which is the split the existing
  xarray-overhead benchmarks use.

`docs/wildfire_applications.md` gains a "Performance and sizing" section with
the measured tables, the reproduction commands, and the memory guidance.

## Headline numbers

Measured on Apple silicon, macOS, CPython 3.14.7, single process, pure NumPy
path, best of three runs:

| Index | Grid | Record | Throughput |
| --- | --- | --- | --- |
| CFFWIS, all seven outputs | 256 x 256 cells | 365 days | 7.4 M cell-days/s |
| CFFWIS, all seven outputs | 1000 x 1000 cells | 30 days | 7.6 M cell-days/s |
| KBDI | 256 x 256 cells | 365 days | 42 M cell-days/s |
| KBDI | 1000 x 1000 cells | 30 days | 43 M cell-days/s |

CONUS at 0.25 degree over 40 years is about 346 M cell-days, so roughly a minute
of CFFWIS recurrence compute and about 8 seconds of KBDI. Memory, not time, is
the binding constraint there: each retained output keeps a full daily history,
so the default seven outputs plus four inputs come to about 30 GB of retained
data at that extent. Measured peak RSS runs about 2 to 2.5 times the modeled
retained footprint (5.3-6.8 GB at 128 x 128 cells over 1825 days, 11.4-12.2 GB
over 3650 days) because the path holds copies of the inputs and outputs
alongside the histories. Peak memory grows with the record length, and no
spatial chunk size changes that.

Both the throughput extrapolation and the memory model are labelled as such in
the docs.

## Measurement correction worth reviewing

The first revision of this branch measured peak RSS as an in-process delta from
a `_PeakRSSMonitor` baseline, reusing the monitor from
`tests/test_benchmark_memory.py`. That metric is order-dependent: an earlier
test in the same session leaves RSS at its high-water mark, so later rows
understate their own run and the `outputs=` pair reported 4.6 GB against 1.7 GB
purely from ordering. Each row is now measured in a fresh process, subtracting
an import-only baseline where the guard compares against the model. The
corrected measurement shows `outputs=` cuts the modeled retained data but not
the measured peak (6.8 GB against 6.3 GB at 1825 days, no reduction at 3650
days), and the docs say that rather than the earlier claim.

## Scope decisions

- **`numba` comparison (task 2) deferred, not dropped.**
  `docs/adr/0006-fire-recursive-state-and-execution.md` already decided that
  `numba` is not an optional dependency and is only reconsidered when a
  representative benchmark misses an agreed runtime target. The ADR keeps that
  decision record in #795, and the throughput table above is the evidence, so
  the accelerator comparison has nothing to compare against. A `numba`
  prototype, if ever needed, belongs in its own ticket behind a runtime target
  that does not exist today.
- **No committed `pytest-benchmark` baseline.** `benchmarks.yml` uploads
  `benchmark-results.json` as an artifact and skips the baseline comparison when
  `.benchmarks/baseline.json` is absent, which is the current state of `main`.
  Committing runner-dependent timing as a baseline would make CI compare numbers
  it cannot compare, so the regression guard is the ratio/RSS budget above
  instead: the results are committed and reproducible as the published tables
  plus the module and the commands that regenerate them.
- Follow-ups #811 (ERA5 CONUS notebook) and #813 (scope boundary) remain the
  open wave-5 work in epic #793.

## Validation

```
uv run ruff check src/ tests/
uv run ruff format --check src/ tests/
uv run mypy src/
uv run pytest                              # 2062 passed
uv run pytest tests/test_benchmark_fire.py -m benchmark -k guards
uv run pytest tests/test_review_scripts.py tests/test_release_integrity.py
uv run scripts/generate_llms_txt.py        # docs/architecture.md is a bundle source
```

With `psutil` unimportable, the six `TestFireBudgetPolicy` tests still run (only
the memory benchmarks skip), which is what the minimum-dependency CI job needs.

Sizing runs used `BENCHMARK_FIRE_GRID_SIDES=256 BENCHMARK_FIRE_RECORD_DAYS=365`,
`BENCHMARK_FIRE_GRID_SIDES=1000 BENCHMARK_FIRE_RECORD_DAYS=30`,
`BENCHMARK_FIRE_MEMORY_GRID_SIDE=128 BENCHMARK_FIRE_MEMORY_RECORD_DAYS=1825
BENCHMARK_FIRE_CHUNK_SIDES=32,64,128`, and a 3650-day memory run at chunk 64.

## Summary by Sourcery

Add fire-index performance benchmarks and publish the resulting runtime and memory sizing guidance with pull-request regression guards.

New Features:
- Add fire-index benchmarks covering throughput, orchestrator cost, and peak memory across configurable workloads.

Enhancements:
- Add deterministic regression budget guards for recurrence speed, orchestrator overhead, and memory usage.
- Publish fire-index performance measurements, sizing guidance, extrapolations, and execution constraints for large gridded workloads.

CI:
- Run fire benchmark and regression guards in the benchmarks workflow on every pull request.

Documentation:
- Document recursive fire-index performance and memory sizing guidance in the wildfire applications documentation.
- Update the architecture test inventory to include fire benchmarks.

Tests:
- Add deterministic tests for the fire benchmark budget policies and failure diagnostics.

### CODE REVIEW

File: tests/test_benchmark_fire.py [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+        """Verify the single-pass orchestrator costs no more than chained calls."""
+        weather = _weather_arrays(_GUARD_RECORD_DAYS, _GUARD_GRID_SIDE**2)
+        single_pass = _measure(lambda: _single_pass_cffwis(weather))
+        chained = _measure(lambda: _chained_cffwis(weather))
+
+        _assert_ratio_within_budget(
+            "single-pass CFFWIS",
+            single_pass,
+            chained,
+            _ORCHESTRATOR_RATIO_BUDGET,
```
[the remark this answers is not in the document]
Comment (monocongo) [reply tests/test_benchmark_fire.py:424]: Fixed in d283d0ec. The enforced budget is unchanged at 1.25x, and the docstrings, the budget comment, and `docs/wildfire_applications.md` now state that permitted cost ("more than 1.25x the chained calls") instead of claiming any slowdown fails. Headroom over the measured 0.86-0.91 ratio is kept deliberately for runner variance.

File: tests/test_benchmark_fire.py [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+            reference_seconds,
+            _MACHINE_SPEED_RATIO_BUDGET,
+            "the recurrence slowed down, or the index gained per-day work",
+        )
+
+    def test_chunking_holds_peak_memory_near_the_model(self) -> None:
+        """Verify chunked CFFWIS peak RSS stays within reach of the modeled footprint."""
+        n_days = _MEMORY_RECORD_DAYS
+        n_side = _MEMORY_GRID_SIDE
+        inputs = _chunked_cffwis_inputs(n_days, n_side, _CHUNK_SIDES[0])
```
[the remark this answers is not in the document]
Comment (monocongo) [reply tests/test_benchmark_fire.py]: Fixed in d283d0ec. `test_chunking_holds_peak_memory_near_the_model` now builds `_chunked_cffwis_inputs(...)` inside the `_PeakRSSMonitor` block, so the measured interval covers the four retained input histories that `_CFFWIS_INPUTS` already counts in the modeled footprint. The guard was re-run with `--benchmark-enable` and passes.

File: docs/wildfire_applications.md [hunk 3]
```diff
@@ -126,14 +126,76 @@ wind is doing. Both are computed from standard reanalysis fields, HDW from a
...
+3.14.7, single process, pure NumPy path, best of three runs):
+
+| Index | Grid | Record | Throughput |
+| --- | --- | --- | --- |
+| CFFWIS, all seven outputs | 256 x 256 cells | 365 days | 7.1 M cell-days/s |
+| CFFWIS, all seven outputs | 1000 x 1000 cells | 30 days | 6.9-7.4 M cell-days/s |
+| KBDI | 256 x 256 cells | 365 days | 41 M cell-days/s |
+| KBDI | 1000 x 1000 cells | 30 days | 42 M cell-days/s |
+
+Reproduce the figures with `pytest tests/test_benchmark_fire.py -m benchmark
```
[the remark this answers is not in the document]
Comment (monocongo) [reply docs/wildfire_applications.md]: Fixed in d283d0ec. The reproduction guidance now names the complete settings per published row: `FIRE_BENCH_MEMORY_GRID_SIDE=128`, `FIRE_BENCH_MEMORY_RECORD_DAYS=1825`, `FIRE_BENCH_CHUNK_SIDES=32,64,128` for the chunk rows, and `FIRE_BENCH_CHUNK_SIDES=64` for the `outputs=("fwi",)` row, which runs at the last configured chunk side.

**Comment by monocongo** [comment]
### 🛠️ Multi-Agent Review Summary
- **Review target:** https://github.com/monocongo/climate_indices/pull/972 (`perf/812-fire-benchmarks`, base `main`)
- **Lanes run:** 5 dispatches, 3 personas returning findings: performance (`lane-performance`), tests (`lane-tests-ds`), maintainability (`lane-maintainability-ds`). Two first-choice lanes failed on provider errors and were retried on alternate models: `lane-tests` (usage limit), `lane-maintainability` (model unsupported). No lane failed to produce a result after retry.
- **Validation:** `uv run ruff check src/ tests/`, `uv run ruff format --check src/ tests/`, `uv run mypy src/` (28 files), `uv run pytest` (2062 passed), `uv run pytest tests/test_benchmark_fire.py -m benchmark -k guards` (3 passed), `uv run pytest tests/test_review_scripts.py tests/test_release_integrity.py` (45 passed), `uv run scripts/generate_llms_txt.py` (bundle regenerated), plus a blocked-psutil run confirming the six policy tests still execute.
- **Fix commits:** [7ca34ae2 — test(fire): measure peak RSS per process and tighten the benchmark budgets](https://github.com/monocongo/climate_indices/commit/7ca34ae2), building on [d283d0ec — test(fire): align benchmark guards and reproduction guidance](https://github.com/monocongo/climate_indices/commit/d283d0ec)
- **Note on branch authorship:** while the lanes ran, `d283d0ec` and a merge of `main` (`163f0bcf`) were committed and pushed to this branch from outside this session. Both were reviewed in full, kept rather than rewritten, and superseded where the fixes below go further.

### Findings and fix status
#### [P2] Benchmark module was skipped entirely without `psutil` — Fixed
- **Location:** `tests/test_benchmark_fire.py` (module import), `tests/test_benchmark_memory.py:59`, `pyproject.toml:42-51`, `.github/workflows/unit-tests-workflow.yml:168`
- **Origin:** Tests
- **Problem:** importing `_PeakRSSMonitor` from `tests/test_benchmark_memory.py` executed that module's `pytest.importorskip("psutil")`, so the minimum-dependency CI job (`uv sync --no-dev --group test`, which installs no psutil) skipped this whole module, including the six `TestFireBudgetPolicy` tests that need no psutil — the module's claimed default-suite coverage.
- **Applied fix:** the monitor import is gone; peak RSS is now measured in a fresh process via `resource.getrusage`, which needs no psutil. Verified by running with psutil blocked: 6 policy tests pass, the 5 memory tests skip cleanly. Commit `7ca34ae2`.

#### [P2] Machine-speed guard could not catch the regression class it claimed — Fixed
- **Location:** `tests/test_benchmark_fire.py` (`_MACHINE_SPEED_RATIO_BUDGET`)
- **Origin:** Performance, Tests
- **Problem:** budget 40.0 against a measured 11.6-14.7 meant a change that doubled the recurrence's per-cell-day cost still passed, while the docstring claimed the guard "fails on a deliberate slowdown".
- **Applied fix:** budget tightened to 20.0, set from five measured runs (11.27-11.76) rather than one sample, and the policy test now asserts arithmetically that a doubling trips the budget. Commit `7ca34ae2`.

#### [P2] Orchestrator guard permitted 25% more cost than its claim — Fixed
- **Location:** `tests/test_benchmark_fire.py` (`_ORCHESTRATOR_RATIO_BUDGET`), `docs/wildfire_applications.md`
- **Origin:** Performance, Maintainability
- **Problem:** the enforced budget was 1.25 while the docstrings and docs said the guard fails when the orchestrator "costs more than the chained calls it replaces", so a 20-30% regression passed.
- **Applied fix:** budget tightened to 1.10 against five measured runs (0.898-0.907), and the docs and docstrings now state the permitted cost instead of the unqualified claim. Wording first aligned in `d283d0ec`, budget tightened in `7ca34ae2`.

#### [P2] Published peak-RSS numbers came from an order-dependent metric — Fixed
- **Location:** `tests/test_benchmark_fire.py` (memory benchmarks and guard), `docs/wildfire_applications.md`
- **Origin:** Performance, Maintainability
- **Problem:** an in-process `_PeakRSSMonitor` delta is measured against a baseline taken after the previous test already raised RSS, so later rows understated their own run: the `outputs=` pair reported 4.6 GB against 1.7 GB, and in a different ordering both reported 0 MB. Published numbers and the `outputs=` guidance were therefore not reproducible.
- **Applied fix:** each row is measured in a fresh process (`resource.getrusage`), with an import-only baseline subtracted where the guard compares against the model. The corrected measurement overturns the earlier claim: `outputs=` cuts the modeled retained data but not the measured peak (6.8 GB versus 6.3 GB at 1825 days, no reduction at 3650 days). Docs table and prose rewritten to the measured values (5.3-6.8 GB at 1825 days, 11.4-12.2 GB at 3650 days, about 2-2.5x the modeled footprint), and the PR body corrected. Commit `7ca34ae2`.

#### [P2] Documented reproduction steps could not reproduce the published tables — Fixed
- **Location:** `docs/wildfire_applications.md`, `tests/test_benchmark_fire.py` (env-var docstring)
- **Origin:** Performance, Maintainability
- **Problem:** the grid benchmarks run at the last configured record length, and the memory rows come from separate env vars, so the stated command produced configurations absent from the tables; the docstring's "spec" values matched no published row.
- **Applied fix:** docs now name the env vars per table and per row, the module docstring lists the three runs behind the published tables, and the env vars follow the existing `BENCHMARK_*` convention (`BENCHMARK_FIRE_*`). Commit `7ca34ae2`.

#### [P2] CONUS extrapolation stated without its grid assumption — Fixed
- **Location:** `docs/wildfire_applications.md`
- **Origin:** Performance
- **Problem:** "roughly 100 x 237 cells" appeared with no lat/lon bounds or source, and the linearity claim rested on two rows plus unstated small-grid overheads.
- **Applied fix:** the assumed bounds (25-50 N, 125-66 W at 0.25 degrees), the 14,610-day record, and the resulting 346 M cell-days are stated; the extrapolation is labelled as such, and the small-grid regime (1.2-5.7 M cell-days/s at 8x8 to 32x32 cells, where per-day call overhead dominates) is documented so the linear model is not over-read. Commit `7ca34ae2`.

#### [P2] `outputs=` guidance overstated what is dropped — Fixed
- **Location:** `docs/wildfire_applications.md`, `tests/test_benchmark_fire.py` module docstring
- **Origin:** Performance
- **Problem:** `outputs=("fwi",)` still retains the FFMC, DMC, and DC histories because the derived indices read them, so "drops the histories of the fields a caller does not need" reads as retaining a single field.
- **Applied fix:** docs state what is retained and what the measured peak does and does not show. Commit `7ca34ae2`.

#### [P3] Memory guard slack was loose and uncalibrated — Fixed
- **Location:** `tests/test_benchmark_fire.py` (`_MEMORY_SLACK`)
- **Origin:** Performance, Tests
- **Problem:** a 3.0 slack on a model that did not match the measured interval was both permissive and mis-specified; at default CI scale the guard could not fire at all.
- **Applied fix:** the interval and the model now cover the same allocations, the slack is 4.0 with the observed 1.9-2.5x range documented as its calibration, and the module docstring states that the guard is a coarse smoke test at CI scale and a real bound at the published settings. Commit `7ca34ae2`.

#### [P3] Guard docstring described a boundary the assertion does not enforce — Fixed
- **Location:** `tests/test_benchmark_fire.py::TestFireBudgetPolicy::test_rejects_recurrence_slowdown_past_budget`
- **Origin:** Maintainability
- **Problem:** "at the boundary included" contradicted the `<=` comparison, where the boundary passes.
- **Applied fix:** renamed and reworded to "just past the budget", with the arithmetic doubling case asserted in the same test. Commit `7ca34ae2`.

#### [P3] Environment parsing and chunk validation were fragile — Fixed
- **Location:** `tests/test_benchmark_fire.py` (`_env_ints`, `_chunked_cffwis_inputs`)
- **Origin:** Tests
- **Problem:** an empty or trailing-comma env value raised `ValueError` during collection for the whole module, and a chunk side that does not divide the grid surfaced as an opaque Dask `ValueError` mid-test.
- **Applied fix:** `_env_ints` filters empty parts and raises a named error when no integer remains; `_chunked_cffwis_inputs` validates divisibility with a readable message. Commit `7ca34ae2`.

#### [P3] Architecture benchmark list omitted the new module — Fixed
- **Location:** `docs/architecture.md:421-424`
- **Origin:** Maintainability
- **Problem:** the benchmark test tree did not list `test_benchmark_fire.py`, drifting from the repository layout it documents.
- **Applied fix:** added the entry and regenerated `llms-full.txt`, since `docs/architecture.md` is in the generator's `FULL_FILES`. Commit `7ca34ae2`.

#### [P4] `FIRE_BENCH_*` env names broke sibling convention — Fixed
- **Location:** `tests/test_benchmark_fire.py`
- **Origin:** Maintainability
- **Problem:** the neighbouring benchmark modules use `BENCHMARK_*`.
- **Applied fix:** renamed to `BENCHMARK_FIRE_*` across the module and docs. Commit `7ca34ae2`.

### Verified clean
- **Docs build path:** the MyST concern is a non-issue: `docs/wildfire_applications.rst` includes the Markdown through `:parser: myst_parser.parsers.docutils_`.
- **Issue references:** #811 is the ERA5 CONUS notebook, #813 the NFDRS/sibling-repo scope decision, #812 this ticket, #795 the design ticket ADR-0006 points at for the NumPy-versus-numba record.
- **llms bundle gate:** `docs/wildfire_applications.md` is not a bundle source, so only the `docs/architecture.md` edit required regeneration; `tests/test_review_scripts.py` passes with the regenerated files.

### Remaining blockers
None. No P0/P1 findings; every verified P2-P4 finding is fixed and re-validated on the current head.

**Comment by monocongo** [comment]
### 🛠️ Multi-Agent Review Summary
- **Review target:** https://github.com/monocongo/climate_indices/pull/972 — merged as 75590cfa; final content head 7ca34ae2
- **Lanes run:** 7 review-lane runs across both heads — correctness/dataflow, benchmark performance ×2, tests, maintainability/docs ×2, security (`deepseek/deepseek-v4-pro`; two initial lanes failed on the exhausted Codex quota and were retried on DeepSeek). Fix commits were re-validated by two further independent lanes.
- **Validation:** `uv run ruff check src/ tests/` pass; `uv run ruff format --check src/ tests/` pass; `uv run mypy src/` pass; `uv run pytest -q` 2066 passed, 423 deselected; `uv run pytest tests/test_benchmark_fire.py -m benchmark --benchmark-enable -k guards` 3 passed; default-suite policy tests 7 passed; psutil-absent simulation on the original head passed (policy tests run, only RSS tests skip); benchmark CI job passed on ubuntu for 7ca34ae2
- **Fix commits:** 7ca34ae2 test(fire): measure peak RSS per process and tighten the benchmark budgets (merged); follow-up PR #974 — ae156dc4 and e3180f2e (docs/naming corrections and probe diagnostics). A parallel local fix 17c75005 was superseded by 7ca34ae2 and never pushed.

### Findings and fix status

#### [P2] Whole benchmark module silently skipped without `psutil` — Fixed (merged, 7ca34ae2)
- **Location:** `tests/test_benchmark_fire.py` (original module-scope monitor import, formerly line 56)
- **Origin:** tests + security + maintainability lanes
- **Problem:** importing `tests.test_benchmark_memory` at module scope executed its `pytest.importorskip("psutil")` during collection, so all of `tests/test_benchmark_fire.py` — including `TestFireBudgetPolicy`, which the docs say the default suite runs, and the CI guards — reported skipped when psutil was unavailable. psutil is only in the `performance` extra, so the `test-minimum-deps` job was exactly that environment: the guards failed open.
- **Applied fix:** the merged rewrite measures peak RSS in fresh subprocesses and no longer imports psutil or the monitor module at all; the default-suite policy tests run in every job.

#### [P3] The published throughput table could not be reproduced with the documented command — Fixed (merged, 7ca34ae2)
- **Location:** `docs/wildfire_applications.md` (reproduction guidance)
- **Origin:** maintainability + performance lanes
- **Problem:** the docs gave one env-var pair for a table whose two configurations need `256,365` and `1000,30` runs.
- **Applied fix:** the merged docs state the per-row env values, and the module docstring lists the published runs.

#### [P3] Module docstring "spec" values contradicted the published tables — Fixed (merged, 7ca34ae2)
- **Location:** `tests/test_benchmark_fire.py` (env-var documentation)
- **Origin:** performance + maintainability lanes
- **Problem:** the docstring claimed `RECORD_DAYS=14610`, grid 256 only, and chunk 64; no published row used those.
- **Applied fix:** the merged rewrite renamed the env vars to `BENCHMARK_FIRE_*` and documents CI defaults plus the published run configurations.

#### [P3] Published memory runs and the output-selection comparison were incomplete/misattributed — Fixed (follow-up #974, ae156dc4)
- **Location:** `tests/test_benchmark_fire.py` (run list), `docs/wildfire_applications.md` (memory reproduction text)
- **Origin:** maintainability lane on the rewritten head
- **Problem:** the docstring's run list could not produce the 3650-day row or the 6.8 GB / 6.3 GB output-selection pair, and the pair was attributed to the run whose last chunk side is 128 while 6.8 GB is the 64 x 64 row.
- **Applied fix:** ae156dc4 lists the 3650-day run and the selection run at `BENCHMARK_FIRE_CHUNK_SIDES=64`; e3180f2e names the run behind the small-grid comparison.

#### [P4] Numeric and wording drift in the sizing section — Fixed (follow-up #974, ae156dc4)
- **Location:** `docs/wildfire_applications.md` lines 166–202; `tests/test_benchmark_fire.py` docstring and slack comment
- **Origin:** maintainability lane on the rewritten head
- **Problem:** "(29.9 M cell-days)" held only for the 1825-day rows; "about 2 to 2.5 times" contradicted the table's 2.6x; the CONUS figure said "approximately one minute" for 47 seconds; the `numba` paragraph implied an agreed runtime target existed.
- **Applied fix:** ae156dc4 corrects all four and aligns the docstring/comment multipliers.

#### [P4] Peak-RSS probe hid child stderr on failure — Fixed (follow-up #974, ae156dc4)
- **Location:** `tests/test_benchmark_fire.py` `_measure_peak_rss_mb`
- **Origin:** performance lane on the rewritten head
- **Problem:** `subprocess.run(..., check=True)` surfaced only `CalledProcessError: ... returned non-zero exit status`; the child traceback was captured and discarded, making probe failures hard to diagnose.
- **Applied fix:** ae156dc4 re-raises `RuntimeError` containing the child's stderr, with a deterministic `TestFireBudgetPolicy` test for the failure path.

### Remaining blockers
None. No P0/P1 findings. The remaining P3/P4 findings landed in follow-up PR #974 because #972 merged before the fix pass could push to its branch.
